### PR TITLE
Particles

### LINKED ISSUE #2349: Add support for particle visualization
Add support for visualizing Lagrangian particle data

### PR DESCRIPTION
**Pull request by StasJ**
Fix #2349

You can generate test data with `vapor/share/examples/generateExampleDCP.py`

# DCP Particle Data Format

**D**ata **C**ollection **P**articles is a simple data format built upon NetCDF for importing particle data into VAPoR.

### P
P is on only required dimension. The size of the dimension P represents the number of particles at this timestep. If the number of particles varies between timesteps, every timestep must be in its own file since each NetCDF file can only define a single length for a dimension.

### T
T is the time dimension and it is required for time-varying data. It should be set to *unlimited*.
### axis
This optional dimension is must have a length of 3 and it is used to pack 3 component values such as position or velocity into a single variable for convenience.

## Variables

The only required variable is Position. Non-time-varying data should have a single dimension of `P`. If the data is time-varying, it should have the dimensions `(T, P)`. 3D particles will often have 3-component data associated with them, for example their velocity would have 3 components, one for each dimension. Since this is a common use case, variables can have 3 components per particle and this is done by setting the fastest-varying dimension to `axis` which as a size of 3. For an example, look at the position variable description below.

### Position

This variable contains the position data for the particles. The data can be stored in either of the below methods:
```
float Position(T, P, axis) ;
```
```
float Position_x(T, P) ;
float Position_y(T, P) ;
float Position_z(T, P) ;
```

### Python Script

`vapor/share/examples/generateExampleDCP.py` shows how to generate a DCP dataset from a particle simulation. 

### Basic NetCDF File

This is an example of a NetCDF file containing 2 particles with time-varying position and particle speed data.

```
netcdf particles_000 {
dimensions:
	P = 2 ;
	T = UNLIMITED ; // (1 currently)
	axis = 3 ;
variables:
	double T(T) ;
		T:units = "seconds" ;
	float Position(T, P, axis) ;
	float speed(T, P) ;
data:

 T = 0 ;

 Position =
  9.762701, 43.03787, 20.55268,
  8.976637, -15.26904, 29.17882 ;

 speed =
  26.02757, 18.87516 ;
}
```

### CODE REVIEW

**Comment by StasJ** [comment]
@sgpearse I tried to run the clang-format script but it is not doing anything.

**Comment by clyne** [comment]
This isn't building for me:

/Users/clyne/src/vapor/apps/vdcdump/vdcdump.cpp:302:95: error: too few arguments to function call,
      expected 3, have 2
        for (int i = 0; i < dimNames.size(); i++) vdc.GetDimension(dimNames[i], _dimensions[i]);
                                                  ~~~~~~~~~~~~~~~~                            ^

File: apps/vaporgui/ParticleEventRouter.h
Comment (clyne) [inline apps/vaporgui/ParticleEventRouter.h:16]: Should Supports{2D,3D} be returning true? This methods refers to a variable's **topolological** (Euclidean) dimension: E.g. zero for points, one for lines, two for planes, and so on.

File: include/vapor/DC.h
Comment (clyne) [inline include/vapor/DC.h:2121]: ts in unused. How come?

File: include/vapor/DataMgr.h
Comment (clyne) [inline include/vapor/DataMgr.h:158]: Wouldn't it make more sense to give ts a default value (-1) so that we don't have to change all of the users of this class? Ditto for some of the other classes for which at 'ts' argument was added?

Comment (StasJ) [reply include/vapor/DataMgr.h:158]: For most of the functions I changed, I did make the ts argument optional (although by overloading rather than a default value, e.g. your comment above about ts being unused). In the case of this function, I determined it was important for the callee to be explicitly aware of whether they pass in the correct timestep since this function is often used to present information to the user/get information about the datasets outside of the internal systems which would already be aware of whether they need the timestep information. It would therefore make it very easy for end users to call this function without being aware that they might need timestep information which is often the case, now that we've added this capability.

File: include/vapor/DC.h
Comment (StasJ) [reply include/vapor/DC.h:2121]: DCs do not support time-varying dimensions by default.

File: apps/vaporgui/ParticleEventRouter.h
Comment (StasJ) [reply apps/vaporgui/ParticleEventRouter.h]: Fixed.

**Comment by StasJ** [comment]
> This isn't building for me:
> 
> /Users/clyne/src/vapor/apps/vdcdump/vdcdump.cpp:302:95: error: too few arguments to function call,
> expected 3, have 2
> for (int i = 0; i < dimNames.size(); i++) vdc.GetDimension(dimNames[i], _dimensions[i]);

Fixed.

File: include/vapor/DataMgr.h
Comment (clyne) [reply include/vapor/DataMgr.h:158]: okey doke

File: lib/vdc/UnstructuredGrid3D.cpp
Comment (clyne) [inline lib/vdc/UnstructuredGrid3D.cpp:41]: These should all be the same size, correct? Just a sanity check.

File: lib/vdc/UnstructuredGrid3D.cpp
Comment (clyne) [inline lib/vdc/UnstructuredGrid3D.cpp:143]: How about deleting the rest of the body of this function since it's not being used? If it's a WIP than maybe just ifdef it out.

File: share/examples/generateExampleDCP.py
Comment (clyne) [inline share/examples/generateExampleDCP.py:8]: Can we add command line arguments to control the number of points and time steps? Optionally outputting velocities would also be useful.

File: lib/render/ParticleRenderer.cpp
Comment (clyne) [inline lib/render/ParticleRenderer.cpp:105]: Better to use DataMgrUtils::GetGrids which will fetch multiple variables at once and temporarily "lock" them so that fetching one variable doesn't purge another one from the cache. Alternatively, you can use DataMgr::GetVariable() with lock=true, but then you'll need to explicitly unlock them.

File: lib/vdc/DataMgr.cpp
Comment (clyne) [inline lib/vdc/DataMgr.cpp:648]: Is this just test code? Are we planning on enabling it? Need to handle error returns below if so. Also, it looks like the position_x variable is hardcoded. Is that reasonable?

File: lib/vdc/DerivedParticleDensity.cpp
Comment (clyne) [inline lib/vdc/DerivedParticleDensity.cpp:94]: Can we remove this?

File: lib/vdc/DerivedParticleDensity.cpp
Comment (clyne) [inline lib/vdc/DerivedParticleDensity.cpp:311]: Will this work as anticipated? I.e. does it give the same result as std::numeric_limits<float>::lowest()?

File: lib/vdc/DerivedVar.cpp
Comment (clyne) [inline lib/vdc/DerivedVar.cpp:867]: REmove commented out code

File: lib/vdc/GridHelper.cpp
Comment (clyne) [inline lib/vdc/GridHelper.cpp:650]: Remove commented out code. A quadtree wouldn't work for 3D data anyway.

**Review by clyne — COMMENTED** [review]
Stas, I've finished reviewing the code and left a number of minor in-line comments. A few more general things that may need some discussion:

1. We use the term "dimension" throughout the application pretty sloppily. It can refer to the Topological (Euclidean) dimension of the individual elements making up a mesh (e.g. points, lines, polygons, polyhedra), or the number of indices needed to address a node or cell in a mesh, or the number of coordinate variables. Sometimes these are all the same, but not always. In the case of particle data the Topological dimension is 0; we have a collection of points with no "cells". You've chosen to represent them internally as a 3D unstructured grid that has no connectivity information. This is reasonable, but maybe leads to confusion. Perhaps it should be a 0D unstructured grid (maybe just call it "Points", but still derive from the UnstructuredGrid class). The real issue is that DC::Mesh::GetTopologyDim() returns 3 for the new class, which is not really correct. One implication of this confusion is that all of the renderers that support 3D data (e.g. slice, volume) are activated for particle data (and will crash if  enabled).
2. A minor thing, but the NetCDF particle format is not recognized if you pass the files to the application from the command line.
3. What are you thinking for the "Melanie" data reader? Currently its a hidden feature. Maybe just ifdef it out?
4. Once we resolve (1) we should add the new particle grid class to Scott's 'smoke tests' (gridTools.cpp and testGrid.cpp). This may flush out some bugs.
5. I see there is a .md file for the DCP data format. We need to make sure this is accessible from the documentation on our web site.

File: lib/vdc/GridHelper.cpp
Comment (StasJ) [reply lib/vdc/GridHelper.cpp]: Done.

File: lib/vdc/DerivedVar.cpp
Comment (StasJ) [reply lib/vdc/DerivedVar.cpp]: Done.

File: lib/vdc/DerivedParticleDensity.cpp
Comment (StasJ) [reply lib/vdc/DerivedParticleDensity.cpp:311]: Yes.

File: lib/vdc/DerivedParticleDensity.cpp
Comment (StasJ) [reply lib/vdc/DerivedParticleDensity.cpp:94]: This is part of an in-progress prototype so it has some commented out code that I was hoping to leave but I can remove it if you insist.

File: lib/vdc/DataMgr.cpp
Comment (StasJ) [reply lib/vdc/DataMgr.cpp:648]: This is prototype code for automatically generating particle density fields when users load particle data. It is in a stage where if you set DCP_ENABLE_PARTICLE_DENSITY it will work but it will have a fixed grid size and it uses a sub-optimal sampling algorithm.

File: lib/render/ParticleRenderer.cpp
Comment (StasJ) [reply lib/render/ParticleRenderer.cpp]: Done

File: share/examples/generateExampleDCP.py
Comment (StasJ) [reply share/examples/generateExampleDCP.py:8]: This script is designed to show users how to output DCP data. It is very heavily documented and designed to be edited by the end user. I think it would be reasonable that if someone, for some reason, wants to customize the example dataset they just edit the file. Also, this script is designed to be as simple and self-explanatory as possible and adding a bunch of argument-handling will, I believe, complicate it unnecessarily. If you insist, however, I will add them.

File: lib/vdc/UnstructuredGrid3D.cpp
Comment (StasJ) [reply lib/vdc/UnstructuredGrid3D.cpp:142]: Done

File: lib/vdc/UnstructuredGrid3D.cpp
Comment (StasJ) [reply lib/vdc/UnstructuredGrid3D.cpp:41]: Correct

File: lib/vdc/DataMgr.cpp
Comment (clyne) [reply lib/vdc/DataMgr.cpp:648]: Could you add a comment to that effect?

**Comment by StasJ** [comment]
> Stas, I've finished reviewing the code and left a number of minor in-line comments. A few more general things that may need some discussion:
> 
> 1. We use the term "dimension" throughout the application pretty sloppily. It can refer to the Topological (Euclidean) dimension of the individual elements making up a mesh (e.g. points, lines, polygons, polyhedra), or the number of indices needed to address a node or cell in a mesh, or the number of coordinate variables. Sometimes these are all the same, but not always. In the case of particle data the Topological dimension is 0; we have a collection of points with no "cells". You've chosen to represent them internally as a 3D unstructured grid that has no connectivity information. This is reasonable, but maybe leads to confusion. Perhaps it should be a 0D unstructured grid (maybe just call it "Points", but still derive from the UnstructuredGrid class). The real issue is that DC::Mesh::GetTopologyDim() returns 3 for the new class, which is not really correct. One implication of this confusion is that all of the renderers that support 3D data (e.g. slice, volume) are activated for particle data (and will crash if  enabled).

All of your points here have been addressed. Particle data now has a topological dimension of 0. There is now a distinction between particle data and regular data. The particle renderer is now only available for datasets with particle data and the other renderers for datasets with regular data. The variable selectors also now only show particle or regular data, depending on the renderer. A few spots where the topological and geometric dimensions were used incorrectly have also been fixed since having a 0D dataset caused bugs to surface.

> 2. A minor thing, but the NetCDF particle format is not recognized if you pass the files to the application from the command line.

This has been fixed.

> 3. What are you thinking for the "Melanie" data reader? Currently its a hidden feature. Maybe just ifdef it out?

I ifdef'd out the Melanie data reader.

> 4. Once we resolve (1) we should add the new particle grid class to Scott's 'smoke tests' (gridTools.cpp and testGrid.cpp). This may flush out some bugs.

A large part of the "smoke tests" is testing sampling the grid at user coordinates which the particles grid does not support. We can implement the iterator tests but the node iterators are pretty much no-ops for particle data and the coord iterators are just 3 UnstructuredGridCoordless iterators.

> 5. I see there is a .md file for the DCP data format. We need to make sure this is accessible from the documentation on our web site.

I think this should probably wait until we merge the PR or maybe even until the next release since the feature will be unavailable to most users.

File: lib/vdc/DataMgr.cpp
Comment (StasJ) [reply lib/vdc/DataMgr.cpp:650]: Done.

File: include/vapor/NetCDFCollection.h
Comment (shaomeng) [inline include/vapor/NetCDFCollection.h:788]: do you want to delete this line?

File: include/vapor/UnstructuredGrid3D.h
Comment (shaomeng) [inline include/vapor/UnstructuredGrid3D.h:35]: do you want to delete this unused function ?

File: lib/render/ParticleRenderer.cpp
Comment (shaomeng) [inline lib/render/ParticleRenderer.cpp:168]: do you want to delete the debug code?

File: lib/vdc/DerivedParticleDensity.cpp
Comment (shaomeng) [inline lib/vdc/DerivedParticleDensity.cpp:188]: remove?

File: lib/vdc/DerivedParticleDensity.cpp
Comment (shaomeng) [inline lib/vdc/DerivedParticleDensity.cpp:160]: remove this?

File: lib/vdc/NetCDFCollection.cpp
Comment (shaomeng) [inline lib/vdc/NetCDFCollection.cpp:149]: remove debug code?

File: lib/vdc/UnstructuredGrid3D.cpp
Comment (shaomeng) [inline lib/vdc/UnstructuredGrid3D.cpp:97]: Remove debug code?

**Review by shaomeng — APPROVED** [review]
Minor changes required, otherwise things look good!

**Comment by clyne** [comment]
@sgpearse will open an issue to add user documentation

File: lib/vdc/UnstructuredGrid3D.cpp
Comment (StasJ) [reply lib/vdc/UnstructuredGrid3D.cpp:97]: I would like to leave this for later work.

File: lib/vdc/NetCDFCollection.cpp
Comment (StasJ) [reply lib/vdc/NetCDFCollection.cpp]: Done.

File: lib/vdc/DerivedParticleDensity.cpp
Comment (StasJ) [reply lib/vdc/DerivedParticleDensity.cpp]: Done.

File: lib/vdc/DerivedParticleDensity.cpp
Comment (StasJ) [reply lib/vdc/DerivedParticleDensity.cpp:187]: I would like to leave this for future reference.

File: lib/render/ParticleRenderer.cpp
Comment (StasJ) [reply lib/render/ParticleRenderer.cpp:168]: I would like to leave this for later work.

File: include/vapor/UnstructuredGrid3D.h
Comment (StasJ) [reply include/vapor/UnstructuredGrid3D.h]: Done.

File: include/vapor/NetCDFCollection.h
Comment (StasJ) [reply include/vapor/NetCDFCollection.h]: Done.

File: include/vapor/DCP.h
Comment (clyne) [inline include/vapor/DCP.h:170]: Please follow underscore convention for naming private variables.

**Review by clyne — COMMENTED** [review]
Stas, this is really great! A few comments/questions

1. You implemented a novel particle visualization capability a while back that can be seen at about the 1'30" mark of this [video](https://www.youtube.com/watch?v=tZxH4-FcAqY). 

How much effort would be involved in adding support for your faded particle traces?

2. it would be really beneficial to be able to subsample the data somewhere in the pipeline. This reduces visual clutter, and improves performance. A simple, user-specifiable stride would help a lot. Some kind of random sampling would be even better. I'm not suggesting you implement these, but you might check out "smarter" sampling approaches in the literature such as in [here](https://dl.acm.org/doi/pdf/10.1145/3281464.3281467https://dl.acm.org/doi/pdf/10.1145/3281464.3281467) and references therein. 

3. Rendering performance is not so great. I generated 2M particles and the frame rates on my Mac were pretty slow. What would could be done to improve this? It looks like currently the entire render pipeline executes for each paint event (i.e. there is no caching). FYI some particle data sets can have billions of particles, but order millions is more common. Again, a down sampling scheme is probably critical.

**Comment by StasJ** [comment]
> Stas, this is really great! A few comments/questions
> 
> 1. You implemented a novel particle visualization capability a while back that can be seen at about the 1'30" mark of this [video](https://www.youtube.com/watch?v=tZxH4-FcAqY).
> 
> How much effort would be involved in adding support for your faded particle traces?

There are two different techniques mentioned here:
- The method in the video would be doable, I don't know how much effort exactly but it would not be unreasonable.
- Adding support for faded particle traces would take a large amount of work. This would require tracking particles through time using their particle id which is currently not possible as Vapor doesn't support integer data grids, in addition to the work it would require to implement the rest of the feature.

> 
> 1. it would be really beneficial to be able to subsample the data somewhere in the pipeline. This reduces visual clutter, and improves performance. A simple, user-specifiable stride would help a lot. Some kind of random sampling would be even better. I'm not suggesting you implement these, but you might check out "smarter" sampling approaches in the literature such as in [here](https://dl.acm.org/doi/pdf/10.1145/3281464.3281467https://dl.acm.org/doi/pdf/10.1145/3281464.3281467) and references therein.

This is already in the PR. Under the data fidelity section there is a `Stride` parameter.

> 2. Rendering performance is not so great. I generated 2M particles and the frame rates on my Mac were pretty slow. What would could be done to improve this? It looks like currently the entire render pipeline executes for each paint event (i.e. there is no caching). FYI some particle data sets can have billions of particles, but order millions is more common. Again, a down sampling scheme is probably critical.

I agree, and there is already a downsampling option as mentioned above. I think this falls in the same situation as adding more rendering features. Any and all of these things can be added, it is just a matter of is this feature going to be used enough to warrant the effort of implementing these features. My understanding was that this PR was going to create a basic implementation that would be expanded upon if there is demand.
Of course, if there are any features you would like me to implement, just let me know and I will do so.

File: include/vapor/DCP.h
Comment (StasJ) [reply include/vapor/DCP.h]: Done

**Review by clyne — APPROVED** [review]

**Pull request merged**
